app/src/main/java/protect/card_locker/ManageGroupActivity.java

```diff
@@ -0,0 +1,264 @@
+        String currentGroupName = mGroupNameText.getText().toString();
+        if (!mGroup._id.equals(currentGroupName.trim())) {
+            Group group = mDB.getGroup(currentGroupName.trim());
+            if (group != null) {
+                mGroupNameNotInUse = false;
+                mGroupNameText.setError(getResources().getText(R.string.group_name_already_in_use));
+            } else {
+                mGroupNameNotInUse = true;
+            }
+        }
+    }
+
+    @Override
+    protected void onResume()
+    {
+        super.onResume();
+
+        setTitle(getString(R.string.editGroup, mGroup._id));
+
+        if (mCurrentGroupName == null){
+            mGroupNameText.setText(mGroup._id);
+        }else{
+            mGroupNameText.setText(mCurrentGroupName);
+            checkIfGroupNameIsInUse();
+        }
+
+        mGroupNameText.addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+                mGroupNameText.setError(null);
+                checkIfGroupNameIsInUse();
+            }
+        });
+
+        if (mAdapterState != null){
+            mAdapter.importInGroupState(mAdapterState);
+        }
+        updateLoyaltyCardList();
+    }
+
+    private ArrayList<Integer> adapterStateToIntegerArray(){
+        ArrayList<Integer> ret = new ArrayList<>(mAdapterState.size() * 2);
+        for (Map.Entry<Integer, Boolean> entry : mAdapterState.entrySet()) {
+            ret.add(entry.getKey());
+            ret.add(entry.getValue()?1:0);
+        }
+        return ret;
+    }
+
+    private void integerArrayToAdapterState(ArrayList<Integer> in) {
+        boolean isKey = true;
+        int cardId = 0;
+        mAdapterState = new HashMap<>();
+        for (int value : in) {
+            if (isKey) {
+                cardId = value;
+            } else {
+                mAdapterState.put(cardId, value == 1);
+            }
+            isKey = !isKey;
+        }
+        if(!isKey){
+            throw(new RuntimeException("failed restoring mAdapterState from integer array list"));
+        }
+    }
+
+
+    @Override
+    protected void onSaveInstanceState (@NonNull Bundle outState){
+        super.onSaveInstanceState(outState);
+        if(mAdapterState != null){
+            outState.putIntegerArrayList("mAdapterState", adapterStateToIntegerArray());
+        }
+        if(mCurrentGroupName != null){
+            outState.putString("mCurrentGroupName", mCurrentGroupName);
+        }
+
+    }
+
+    @Override
+    protected void onPause(){
+        super.onPause();
+
+        mAdapterState = mAdapter.exportInGroupState();
+        mCurrentGroupName = mGroupNameText.getText().toString();
+    }
```

Comment:
Why are you using onPause here? In LoyaltyCardEditActivity I am saving right when `onSaveInstanceState` is called. Sounds like you could simplify this by getting rid of the onPause and just writing immediately in onSaveInstanceState? Something like this?

```
    @Override
    protected void onSaveInstanceState (@NonNull Bundle outState){
        super.onSaveInstanceState(outState);
        outState.putIntegerArrayList("mAdapterState", mAdapterState != null ? adapterStateToIntegerArray() : null);
        outState.putString("mCurrentGroupName", mCurrentGroupName);
    }
```

By not only writing when != null you also prevent this value not being emptied correctly in some edge case that may or may not exist.